drv/i2c-devices/src/pca9956b.rs

```diff
@@ -0,0 +1,332 @@
+            } else if err == LedErr::Invalid {
+                summary.invalid += 1;
+            }
+        }
+
+        summary
+    }
+}
+
+/// Auto-increment flag is Bit 7 of the control register. Bits 6..0 are address.
+const CTRL_AUTO_INCR_MASK: u8 = 1 << 7;
+/// The MODE2 OVERTEMP bit indicates if an overtempature condition has occurred
+const MODE2_OVERTEMP_MASK: u8 = 1 << 7;
+/// The MODE2 ERROR bit indicates if any error conditions are in EFLAGn
+const MODE2_ERROR_MASK: u8 = 1 << 6;
+/// The MODE2 CLRERR bit clears all error conditions in EFLAGn
+const MODE2_CLRERR_MASK: u8 = 1 << 4;
+/// The MODE2 reserved bits have a defined pattern (0b101) and are read only.
+/// They will be used to validate a PCA9956B with humility
+const MODE2_RSVD_MASK: u8 = 0x7;
+const MODE2_RSVD: u8 = 0x5;
+
+pub struct Pca9956B {
+    device: I2cDevice,
+}
+
+pub const NUM_LEDS: usize = 24;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    /// The low-level I2C communication returned an error
+    I2cError(ResponseCode),
+
+    /// The LED index is too large
+    InvalidLED(u8),
+}
+
+impl From<ResponseCode> for Error {
+    fn from(err: ResponseCode) -> Self {
+        Error::I2cError(err)
+    }
+}
+
+impl From<Error> for ResponseCode {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::I2cError(code) => code,
+            _ => panic!(),
+        }
+    }
+}
+
+impl Pca9956B {
+    pub fn new(device: &I2cDevice) -> Self {
+        Self { device: *device }
+    }
+
+    /// Read a single Register
+    fn read_reg(&self, reg: Register) -> Result<u8, Error> {
+        self.device
+            .read_reg::<u8, u8>(reg as u8)
+            .map_err(|code| Error::I2cError(code))
+    }
+
+    /// Read a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    #[allow(dead_code)]
+    fn read_buffer(&self, reg: Register, buf: &mut [u8]) -> Result<(), Error> {
+        self.device
+            .read_reg_into((reg as u8) | CTRL_AUTO_INCR_MASK, buf)
+            .map_err(|code| Error::I2cError(code))?;
+
+        Ok(())
+    }
+
+    /// Write a single Register
+    fn write_reg(&self, reg: Register, val: u8) -> Result<(), Error> {
+        let buffer = [reg as u8, val];
+        self.device
+            .write(&buffer)
+            .map_err(|code| Error::I2cError(code))
+    }
+
+    /// Write a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    fn write_buffer(&self, reg: Register, buf: &[u8]) -> Result<(), Error> {
+        let mut data: [u8; 25] = [0; 25];
+        data[0] = (reg as u8) | CTRL_AUTO_INCR_MASK;
+        data[1..=buf.len()].copy_from_slice(buf);
```

Comment:
This will panic if the incoming `buf` is too long.  Where does the `25` limitation come from?

It looks like the longest possible write is `0h-40h`, i.e. 65 bytes.

Should we use that size for `data`, and preemptively return an error if `reg as u8 + buf.len() >= EFLAG0`?

(along with a comment explaining that `EFLAG0` is the first read-only register)

Reply:
Thanks for pointing this out! The 25 was somewhat arbitrary assuming a stride of all PWMx and an address byte. I'll size `data` at `63` (`0h-3Eh`, plus an address byte) and add in the requisite check and comment.